### ISSUE TITLE
Undefined behaviour in `slice::fill` specialization.

### ISSUE THREAD

**Issue by m-ou-se** [rust-lang/rust#87891]
This was reported by @the8472 on Zulip: https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/potential.20UB.20in.20slice.3A.3Afill/near/248871405

> I'm wondering if the optimization here is correct: https://github.com/rust-lang/rust/blob/ae90dcf0207c57c3034f00b07048d63f8b2363c8/library/core/src/slice/specialize.rs#L22-L28
> 
> Specifically if T is `MaybeUninit<u8>`. It momentarily transmutes that to an u8 before passing it to `write_bytes`.

Demo:

```rust
use std::mem::MaybeUninit;

fn main() {
    let mut a = [MaybeUninit::<u8>::uninit(); 10];
    a.fill(MaybeUninit::uninit());
}
```
```
error: Undefined Behavior: using uninitialized data, but this operation requires initialized memory
    --> /home/mara/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/intrinsics.rs:2191:14
     |
2191 |     unsafe { write_bytes(dst, val, count) }
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ using uninitialized data, but this operation requires initialized memory
     |
     = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
     = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
```

**Comment by Aaron1011** [comment]
I think the fix should be as simple as using `MaybeUnint<u8>` instead of `u8` as the type of `value` - that will work regardless of whether or not the memory is initialized.

**Comment by m-ou-se** [comment]
The other option is just removing the `if size_of::<T>() == 1` case entirely. Llvm is smart enough to turn that loop in the `else` branch into a memset.

**Comment by the8472** [comment]
It does help the gcc codgen backend at least as it can't manage to produce a memset for u64 so I assume it would fail to do the same for u8 if it weren't for that specialization.

https://rust.godbolt.org/z/hrMYvWz9n

**Comment by m-ou-se** [comment]
@the8472: It also fails for the `u8` case: https://rust.godbolt.org/z/Wjx5Mnj5K

**Comment by the8472** [comment]
It does call memset if you don't pass an array. https://rust.godbolt.org/z/x4KT99heT

**Comment by RalfJung** [comment]
I'd think the `write_bytes` intrinsic could easily have its type changed to take a `MaybeUninit<u8>`, which would let one solve this problem. The stable function of course cannot be changed, though...

**Comment by RalfJung** [comment]
Reopening to track adding a test. (That's the standard procedure for this, isn't it?)
